scripts/release/check_commit_metadata.sh

```diff
@@ -0,0 +1,108 @@
+#!/usr/bin/env bash
+
+# Usage: source ./check_commit_tags.sh <revision range>
+# Usage: ./check_commit_tags.sh <revision range>
+#
+# Example: ./check_commit_tags.sh v0.13.1..971e3678
+#
+# When sourced, this script will populate the COMMIT_METADATA_* variables
+# with the commit metadata for each commit in the revision range.
+#
+# Because this script does some expensive lookups via the GitHub API, its
+# results will be cached in the environment and restored if this script is
+# sourced a second time with the same arguments.
+
+set -euo pipefail
+# shellcheck source=scripts/lib.sh
+source "$(dirname "${BASH_SOURCE[0]}")/../lib.sh"
+
+range=${1:-}
+
+if [[ -z $range ]]; then
+	error "No revision range specified"
+fi
+
+# Check dependencies.
+dependencies gh
+
+COMMIT_METADATA_BREAKING=0
+declare -A COMMIT_METADATA_TITLE COMMIT_METADATA_CATEGORY
+
+main() {
+	# Match a commit prefix pattern, e.g. feat: or feat(site):.
+	prefix_pattern="^([a-z]+)(\([a-z]*\))?:"
+
+	# If a commit contains this title prefix or the source PR contains the
+	# label, patch releases will not be allowed.
+	# This regex matches both `feat!:` and `feat(site)!:`.
+	breaking_title="^[a-z]+(\([a-z]*\))?!:"
+	breaking_label=release/breaking
+	breaking_category=breaking
+
+	mapfile -t commits < <(git log --no-merges --pretty=format:"%h %s" "$range")
+
+	for commit in "${commits[@]}"; do
+		mapfile -d ' ' -t parts <<<"$commit"
+		commit_sha=${parts[0]}
+		commit_prefix=${parts[1]}
+
+		# Store the commit title for later use.
+		title=${parts[*]:1}
+		title=${title%$'\n'}
+		COMMIT_METADATA_TITLE[$commit_sha]=$title
+
+		# First, check the title for breaking changes. This avoids doing a
+		# GH API request if there's a match.
+		if [[ $commit_prefix =~ $breaking_title ]]; then
+			COMMIT_METADATA_CATEGORY[$commit_sha]=$breaking_category
+			COMMIT_METADATA_BREAKING=1
+			continue
+		fi
+
+		# Get the labels for the PR associated with this commit.
+		mapfile -t labels < <(gh api -H "Accept: application/vnd.github+json" "/repos/coder/coder/commits/${commit_sha}/pulls" -q '.[].labels[].name')
```

Comment:
The rate limit for calls from github actions is 1k per hour per repository. I'd say an upper bound for the amount of PRs in a release would be 100, but combined with other things in the repo we might be pushing on 1k per hour during busy periods. Can this instead use a list endpoints?

The commits we release are a contiguous block on main so if you list from first ref to end ref then you will get them all in much less requests.

Reply:
I did consider this but it's not clear how we'd trigger it. Now I don't remember but do we run dry-run for every merged commit? I considered adding a `--skip-label-check` to this script used in dry-run scenarios, but omitted it for simplicity and because it would generate "false" release notes.

> The commits we release are a contiguous block on main so if you list from first ref to end ref then you will get them all in much less requests.

The problem isn't listing the commits, but in that endpoint they won't include the metadata we're looking for (PR labels). I tried to find a way to get all of this information in one go, but each one has caveats.. 😅

We could use [`/repos/coder/coder/pulls`](https://docs.github.com/en/rest/pulls/pulls?apiVersion=2022-11-28#list-pull-requests), but even with filtering `?state=closed&base=main&per_page=100` (maybe `sort=updated`), we might have to keep traversing multiple pages to find all commits that we're looking for (say we have a month old PR that got merged).

The current API is also interesting in that, I'm speculating here, but it lists PRs _related_ to the commit. As such I'm thinking _maybe_ it would also list a PR that has been merged into a PR that then is merged into `main`. This could be useful.

But alas, that benefit is theoretical and ultimately, in most cases, not beneficial. The pulls endpoint will likely result in less requests so maybe we go with that.

Reply:
> do we run dry-run for every merged commit

We do not

What if you instead listed all PRs that had the breaking label and then checked if the merge commit is from any of those PRs?

Reply:
@deansheather this is now implemented in [`c871cb7` (#5366)](https://github.com/coder/coder/pull/5366/commits/c871cb7744212652fb0c16c3ca5b7b46546f02fe).

I opted to _not_ filter the search results based on a single tag because I wanted to create a safety-check that ensures we fetched PR data for each commit.